base/system/winlogon/rpcserver.c

```diff
@@ -485,6 +507,87 @@ BaseRegQueryMultipleValues(
     return ERROR_SUCCESS;
 }
 
+INT_PTR
+CALLBACK
+ShutdownDialogProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    switch (uMsg)
+    {
+        case WM_INITDIALOG:
+        {
+            g_hShutdownDialog = hwndDlg;
+            if (g_ShutdownParams.usMessage.Length)
+                SetDlgItemTextW(hwndDlg,
+                                IDC_SHUTDOWNCOMMENT,
+                                g_ShutdownParams.usMessage.Buffer);
+            RemoveMenu(GetSystemMenu(hwndDlg, FALSE), SC_CLOSE, MF_BYCOMMAND);
+            SetWindowPos(hwndDlg, HWND_TOPMOST, 0, 0, 0, 0, SWP_NOSIZE | SWP_NOMOVE);
+            PostMessage(hwndDlg, WM_TIMER, 0, 0);
+            SetTimer(hwndDlg, IDT_SYSSHUTDOWN, 1000, NULL);
+        break;
+        }
+        case WM_CLOSE:
+        {
+            g_bShuttingDown = FALSE;
+            KillTimer(hwndDlg, IDT_SYSSHUTDOWN);
+            if (g_ShutdownParams.usMessage.Buffer)
+            {
+                HeapFree(GetProcessHeap(), 0, g_ShutdownParams.usMessage.Buffer);
+            }
+            EndDialog(hwndDlg, 0);
+            DestroyWindow(hwndDlg);
+        break;
+        }
+        case WM_TIMER:
+        {
+            WCHAR strbuf[34];
+            int seconds, minutes, hours;
+            seconds = (int)(g_ShutdownParams.dwTimeout);
+            hours = seconds/3600;
+            seconds -= hours*3600;
+            minutes = seconds/60;
+            seconds -= minutes*60;
+            ZeroMemory(strbuf, sizeof(strbuf));
+            //FIXME: Show time remaining according to the locale's format
+            RtlStringCbPrintfW(strbuf, sizeof(strbuf), L"%d:%d:%d", hours, minutes, seconds);
+            SetDlgItemTextW(hwndDlg, IDC_SHUTDOWNTIMELEFT, strbuf);
+            if (g_ShutdownParams.dwTimeout == 0)
+            {
+                PostMessage(hwndDlg, WM_CLOSE, 0, 0);
+                ExitWindowsEx((g_ShutdownParams.bRebootAfterShutdown ? EWX_REBOOT : EWX_SHUTDOWN) |
+                              (g_ShutdownParams.bForceAppsClosed ? EWX_FORCE : 0),
+                              g_ShutdownParams.dwReason);
+            }
+            g_ShutdownParams.dwTimeout--;
+        break;
+        }
+        default:
+            return FALSE;
+    }
+    return TRUE;
+}
+
+DWORD
+WINAPI
+InitiateSystemShutdownThread(LPVOID lpParameter)
+{
+    INT_PTR status;
+    status = DialogBoxW(hAppInstance, MAKEINTRESOURCEW(IDD_SYSSHUTDOWN),
+                        NULL, ShutdownDialogProc);
+    if (status >= 0)
+    {
+        return ERROR_SUCCESS;
+    }
+    else
+    {
+        if (g_ShutdownParams.usMessage.Buffer)
+        {
+            HeapFree(GetProcessHeap(), 0, g_ShutdownParams.usMessage.Buffer);
```

Comment:
A `RtlInitEmptyUnicodeString(&(g_hShutdownParams.usMessage), NULL, 0);` just after this HeapFree() may be needed to turn the string empty but valid again.